lib/nostr.js

```diff
@@ -16,24 +20,52 @@ export const DEFAULT_CROSSPOSTING_RELAYS = [
   'wss://nostr.mutinywallet.com/',
   'wss://relay.mutinywallet.com/'
 ]
+export const DEFAULT_NIP46_RELAYS = [
+  'wss://relay.damus.io/',
+  'wss://relay.snort.social/'
+]
 
 export class Relay {
-  constructor (relayUrl) {
-    const ws = new WebSocket(relayUrl)
+  constructor (relayUrl, reconnectOnClose = false) {
+    this.url = relayUrl
+    this.subs = []
+    this.reconnectOnClose = reconnectOnClose
+    this._closed = false
+    this.error = undefined
+    this.reconnect()
+  }
 
-    ws.onmessage = function (msg) {
+  reconnect () {
+    this._closed = false
+    const ws = new WebSocket(this.url)
+    this.ws = ws
+
+    ws.onmessage = (msg) => {
       const [type, notice] = JSON.parse(msg.data)
       if (type === 'NOTICE') {
         console.log('relay notice:', notice)
       }
     }
 
-    ws.onerror = function (err) {
+    ws.onerror = (err) => {
       console.error('websocket error: ' + err)
       this.error = err
     }
 
-    this.ws = ws
+    ws.onclose = () => {
+      if (this.reconnectOnClose && !this._closed) {
+        setTimeout(() => {
+          this.reconnect()
+        }, 5000)
+      }
+    }
+
+    ws.onopen = () => {
+      // Resubscribe to all subscriptions on reconnect
+      for (const sub of this.subs) {
+        ws.send(JSON.stringify(['REQ', sub.id, ...sub.filters]))
+      }
+    }
```

Comment:
Mhh, if you add these listeners conditionally, I think we don't need store `this.reconnect` as state. This is the only place you use it so you could consume it immediately when it's passed into `static connect` without any assignment to `this`. For example like this:

```diff
diff --git a/lib/nostr.js b/lib/nostr.js
index 7af160cf..c4b5bbe4 100644
--- a/lib/nostr.js
+++ b/lib/nostr.js
@@ -68,9 +68,15 @@ export class Relay {
     }
   }

-  static async connect (url, { timeout } = {}) {
+  static async connect (url, { reconnect, timeout } = {}) {
     const relay = new Relay(url)
     await relay.waitUntilConnected({ timeout })
+
+    if (reconnect) {
+      // add reconnect logic to websocket
+      // ...
+    }
+
     return relay
   }
```

This might sound petty but I just wrote this code so I'd like to keep it as stateless as possible, lol

Reply:
the ws field should be considered private imo since the relay class is an abstraction over the use of websockets, so it wouldn't be a good practice to 'leak' it outside of the class like so

Reply:
i made a different change : https://github.com/stackernews/stacker.news/pull/1311/commits/a8d572ba452ac01b310fae9e660dfce5cbf9d9c4 to not set this.reconnectOnClose in the constructor, maybe this can be a compromise?